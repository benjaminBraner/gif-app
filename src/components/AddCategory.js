import React, { useState } from 'react';
import PropTypes from 'prop-types';
import '../styles/components/AddCategory.css';

export const AddCategory = ({ setCategories }) => {


     const [inputValue, setInputValue] = useState('');

     const handleInputChange = (e) => {
          setInputValue(e.target.value)
     }

     const handleSubmit = (e) => {
          e.preventDefault();

          if (inputValue.trim().length > 2) {
               setCategories(categories => {
                    let y = [...categories].filter(c => c !== inputValue)
                    return [inputValue, ...y];
               });

               setInputValue('');
          }
     }

     return (
          <>
               <form onSubmit={handleSubmit}>
                    <input
                         type='text'
                         value={inputValue}
                         onChange={handleInputChange}
                    />
                    <button type='submit'>Submit</button>
               </form>
          </>
     )
}

AddCategory.propTypes = {
     setCategories: PropTypes.func.isRequired
}