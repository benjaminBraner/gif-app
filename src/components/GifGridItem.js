import PropTypes from "prop-types";

export const GifGridItem = ({title, url}) => {
   // console.log({id,title,url});
   return (
      <div className="card animate__animated animate__bounce">
         <p>{title}</p>
         <img src={url} alt={title}></img>
      </div>
   )             
}

GifGridItem.propTypes = {
   title: PropTypes.string.isRequired,
   url: PropTypes.string.isRequired
}