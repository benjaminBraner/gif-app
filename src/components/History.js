import '../styles/components/History.css';


export const History = ({ histItem, setCategories }) => {

   const showElement = () => {
      setCategories(categories => {
         let y = [...categories].filter(c => c !== histItem);
         return [histItem, ...y];
      });
   }

   return (
      <>
      {histItem ? <li onClick={showElement}>{histItem}</li> : null}

      
      
      </>
   )
}
