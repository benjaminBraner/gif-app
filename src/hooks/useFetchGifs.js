
import {useState,useEffect} from "react";
import {getGifs} from "../helpers/getGifs";

export const useFetchGifs = (category) => {
   const [state, setState] = useState({
      data: [],
      loading: true
   })

   /*este useEffect me permite ejecutar codigo de manera condicional, como primer argumento reci-
   be una funcion que es el codigo que yo quiero ejecutar, También es posible especificar cuándo 
   se debe ejecutar esta función con un segundo argumento opcional que le podemos pasar.
   Para ello basta con añadir un segundo parámetro a la función, con la lista de los elementos de 
   los que depende. Si el valor de uno de estos elementos que hemos indicado cambia, 
   la función se va a ejecutar con la siguiente actualización.

   Otra posibilidad que nos permite este hook es la de especificar que se ejecute una unica vez,
   solo basta con poner un array vacio como segundo parametro []
   */
   useEffect( () => {
   
      getGifs(category)    
         .then(imgs => {
               setState({
                  data: imgs,
                  loading: false
               })        
         })
   },[category])

   return state; //{data: [], loading: true}
}