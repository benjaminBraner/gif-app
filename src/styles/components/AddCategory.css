input {
   max-width: 500px;
   margin: 10px 5px 10px 10px;
   padding: 6px;
   font-size: 17px;

}

button {
   background-color: rgb(60, 145, 241);
   border-radius: 5px;
   border: none;
   color: #fff;
   padding: 6px;
   margin: 10px 10px 10px 5px;
   font-size: 17px;
   cursor: pointer;
}

button:hover {
   background-color: rgb(17, 97, 189);
}