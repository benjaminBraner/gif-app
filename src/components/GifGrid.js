import { useFetchGifs } from "../hooks/useFetchGifs";
import PropTypes from "prop-types";
import { GifGridItem } from "./GifGridItem";


export const GifGrid = ({ category }) => {
   const { data: images, loading } = useFetchGifs(category);

   return (
      <>
         <h3>{category}</h3>

         {loading && <p>Loading...</p>}
         {/* {loading ? <p>Loading...</p> : null} */}

         <div className="card-grid">
            {
               images.map(img => (
                  <GifGridItem
                     key={img.id}
                     {...img}
                  />
               ))
            }


         </div>
      </>
   )
};

GifGrid.propTypes = {
   category: PropTypes.string.isRequired
}