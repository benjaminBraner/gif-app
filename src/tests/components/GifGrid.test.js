import { shallow } from "enzyme";
import { GifGrid } from "../../components/GifGrid"
import { useFetchGifs } from "../../hooks/useFetchGifs";
jest.mock("../../hooks/useFetchGifs");


describe('Pruebas en el <GifGrid/>', () => {

   const category = "Hello";

   test('debe marcar el snapshot', () => {

      /*lo que hace esto es que cuando se llame el useFetchGifs dentro de mi componente, va a 
      regresar lo que sea que este entre los parentesis, en este caso pongo eso porque por 
      defecto esos son los valores que el useFetchGifs regresa */
      useFetchGifs.mockReturnValue({
         data: [],
         loading: true
      })

      const wrapper = shallow(<GifGrid category={category} />);
      expect(wrapper).toMatchSnapshot();
   })

   test('debe mostrar items cuando se cargan imagenes con el useFetchGifs', () => {


      const gifs = [{
         id: 'abc',
         url: 'https://localhost/hola/imagen.jpg',
         title: 'cualquier cosa'
      },
      {
         id: 'def',
         url: 'https://localhost/hola/imagen.jpg',
         title: 'cualquier cosa'
      }

      ]
      /*a diferencia del anterior test, ahora necesito de alguna manera fingir que ya tengo la data
      de la api, por lo tanto este test es cuando mi useFetchGifs ya me regresa la data
      
      QUE ES UN MOCK?
      es una simulacion manual usada para sustituir funcionalidades con datos falsos, como falsear 
      algo por ejm para que mi componente crea que ya tiene la informacion resuelta, y asi hacer 
      los tests mas rapido
      
      

      antes de que se haga esta renderizacion:
      const wrapper = shallow(<GifGrid category={category}/>);
      
      yo puedo simular que cuando el componente llame a mi useFetchGifs, ya tenga data:
      
      1.importo el useFetchGifs en la parte de arriba
      import { useFetchGifs } from "../../hooks/useFetchGifs";
      jest.mock("../../hooks/useFetchGifs");
      
      el jest.mock("path") es para fingir cualquier llamada a ese archivo y controlar la data que
      este va a responder
      */

      /*ahora lo que hago aqui es lo mismo que el test de arriba pero esta vez evaluandolo cuando
      ya tengo la data, osea ahora voy a suponer que el loading esta en false (es decir que ya
       tengo la informacion) y la data van a ser los gifs*/

      useFetchGifs.mockReturnValue({
         data: gifs,
         loading: false
      })
      const wrapper = shallow(<GifGrid category={category} />);

      //ahora viene la evaluacion

      // expect(wrapper).toMatchSnapshot();
      /*este test funciona pero no es muy confiable porque por ejm, como estoy haciendo el snapshot
      si yo quitara el objeto que esta dentro del array gifs, la prueba va a fallar, pero fallaria
      porque se cambio el snapshot nada mas, y si actualizo el snapshot se resolveria, y no deberia
      de comportarse asi, por eso hay formas mas seguras de evaluar esto de forma mas estricta:*/


      //espero que el parrafo "Loading" no exista porque asi se que ya tiene la data
      expect(wrapper.find('p').exists()).toBe(false);
      /*espero que el numero de componentes GifGridItem (que es el componente que trae cada objeto
      del array gifs por separado) sea igual al numero de objetos que tiene el array gifs*/
      expect(wrapper.find('GifGridItem').length).toBe(gifs.length);

   })

})
