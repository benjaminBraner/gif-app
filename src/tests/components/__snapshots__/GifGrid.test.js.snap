// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el <GifGrid/> debe marcar el snapshot 1`] = `
<Fragment>
  <h3>
    Hello
  </h3>
  <p>
    Loading...
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;
